### PR TITLE
Add jrePath option

### PR DESCRIPTION
**Pull request by petoncle**
This PR implements a new field called jrePath in config.json.
This is my solution to the problem described here: https://github.com/libgdx/packr/issues/191
The main idea is to let the user choose the name of the bundled JRE folder so that a self-updating app can easily switch to a new JRE.
Let me know if this solution is not appropriate and/or if there are better alternatives.

I've also added the --jrePath option to the packaging tool (not the launcher).
The default value is "jre" for backward compatibility.
I've changed the tests to use this new option: "--jrePath jre-123456". The tests are successful on Windows, Linux and Mac.

Anyway this is mostly a proof of concept, I'm trying to get a discussion going about it.

By the way @karlsabo I'm not sure how you build, run the tests and publish launcher natives and packr-all. Are you able to do it in a single step or are there multiple steps involved?
To do it in one single step in my fork I had to make a custom GitHub Actions workflow that 1- publishes the packr launcher native executables to GitHub Packages, and 2- runs the tests.

### CODE REVIEW

File: PackrLauncher/src/main/cpp/win32/packr_win32.cpp [hunk 1]
```diff
@@ -255,7 +256,7 @@ void addDllDirectory(LPCWSTR directory) {
  * Loads every library found using the pattern {@code libraryPattern}.
  * @param libraryPattern the search pattern to use to locate shared libaries to load
  */
-void loadLibraries(const PTCHAR libraryPattern) {
+void loadLibraries(const TCHAR* libraryPattern) {
```
Comment (karlsabo) [inline PackrLauncher/src/main/cpp/win32/packr_win32.cpp:259]: What's the reason for changing from the typedef PTCHAR?

File: PackrLauncher/src/main/headers/packr.h [hunk 2]
```diff
@@ -41,7 +41,7 @@ extern "C" {
 	extern const char __CLASS_PATH_DELIM;
 
 	/* platform-dependent functions */
-	bool loadJNIFunctions(GetDefaultJavaVMInitArgs* getDefaultJavaVMInitArgs, CreateJavaVM* createJavaVM);
+	bool loadJNIFunctions(const std::string& jrePath, GetDefaultJavaVMInitArgs* getDefaultJavaVMInitArgs, CreateJavaVM* createJavaVM);
```
Comment (karlsabo) [inline PackrLauncher/src/main/headers/packr.h:44]: For consistency I would prefer const dropt_char* instead of string.

It would also feel less awkward with the Windows code assuming it's getting a string with UTF-8 encoded data in it.

File: PackrLauncher/src/main/cpp/win32/packr_win32.cpp [hunk 3]
```diff
@@ -282,15 +283,35 @@ void loadLibraries(const PTCHAR libraryPattern) {
...
-bool loadJNIFunctions(GetDefaultJavaVMInitArgs *getDefaultJavaVMInitArgs, CreateJavaVM *createJavaVM) {
-   addDllDirectory(TEXT("jre\\bin"));
-   addDllDirectory(TEXT("jre\\bin\\server"));
+LPCWSTR stringToLpcwstr(const std::string& s) {
+   std::wstring stemp = std::wstring(s.begin(), s.end());
+   LPCWSTR sw = stemp.c_str();
+   return sw;
+}
+
+const TCHAR* stringToTcharPointer(const std::string& s) {
```
Comment (karlsabo) [inline PackrLauncher/src/main/cpp/win32/packr_win32.cpp:292]: We should use PTCHAR for consistency.

File: PackrLauncher/src/main/cpp/packr.cpp [hunk 4]
```diff
@@ -497,7 +497,15 @@ void launchJavaVM(const LaunchJavaVMCallback &callback) {
     GetDefaultJavaVMInitArgs getDefaultJavaVMInitArgs = nullptr;
     CreateJavaVM createJavaVM = nullptr;
 
-    if (!loadJNIFunctions(&getDefaultJavaVMInitArgs, &createJavaVM)) {
+    const string originalJrePath = hasJsonValue(jsonRoot, "jrePath", sajson::TYPE_STRING) ?
+        getJsonValue(jsonRoot, "jrePath").as_string() : "jre";
+    string trimmedJrePath = originalJrePath;
+    // Removes trailing slash.
+    if ('/' == trimmedJrePath.back())
```
Comment (karlsabo) [inline PackrLauncher/src/main/cpp/packr.cpp:504]: I'm assuming a trailing '/' breaks one or more of the loadJNIFunctions implementations? We should also check for '\\' and remove that too, or make a comment if this only affects Linux or macOS.

File: PackrAllTestApp/packrAllTestApp.gradle.kts [hunk 5]
```diff
@@ -287,6 +287,9 @@ fun createPackrContent(jdkPath: Path, osFamily: String, destination: Path) {
       args(jdkPath.toFile())
       args("--executable")
       args("PackrAllTestApp")
+      args("--jrePath")
+      // Tests a custom jre path.
+      args("jre-123456")
```
Comment (karlsabo) [inline PackrAllTestApp/packrAllTestApp.gradle.kts:292]: Thanks for updating the test! Could you add a Unicode character here so we know if it works or not on Windows? You could use Ä, like I do above `// Add the Ä to test unicode paths`

File: PackrLauncher/src/main/cpp/win32/packr_win32.cpp [hunk 6]
```diff
@@ -282,15 +283,35 @@ void loadLibraries(const PTCHAR libraryPattern) {
...
+        pstring = wstr.data();
+    #else
+        pstring = s.data();
+    #endif
+    return pstring;
+}
+
+bool loadJNIFunctions(const std::string& jrePath, GetDefaultJavaVMInitArgs *getDefaultJavaVMInitArgs, CreateJavaVM *createJavaVM) {
+   std::string backslashedJrePath = jrePath;
+   std::replace(backslashedJrePath.begin(), backslashedJrePath.end(), '/', '\\');
```
Comment (karlsabo) [inline PackrLauncher/src/main/cpp/win32/packr_win32.cpp:306]: I don't know for sure, but I could imagine a case where this method segfaults when it's passed a string that actually contains UTF-8 data. I would like this method to take a PTCHAR* (const dropt_char*), then we know what kind of data we're really dealing with.

File: PackrLauncher/src/main/cpp/packr.cpp [hunk 7]
```diff
@@ -497,7 +497,15 @@ void launchJavaVM(const LaunchJavaVMCallback &callback) {
     GetDefaultJavaVMInitArgs getDefaultJavaVMInitArgs = nullptr;
     CreateJavaVM createJavaVM = nullptr;
 
-    if (!loadJNIFunctions(&getDefaultJavaVMInitArgs, &createJavaVM)) {
+    const string originalJrePath = hasJsonValue(jsonRoot, "jrePath", sajson::TYPE_STRING) ?
+        getJsonValue(jsonRoot, "jrePath").as_string() : "jre";
```
Comment (karlsabo) [inline PackrLauncher/src/main/cpp/packr.cpp:501]: Great job handling the backwards compatibility case, it's easy to break these since there aren't any tests for them.

**Review by karlsabo — CHANGES_REQUESTED** [review]
Thanks so much for the PR! I made a few comments I would like to see changed. The biggest one being the parameter type to loadJNIFunctions.

**Comment by karlsabo** [comment]
Oh, please also update the changelog.md, add something like:

# Unreleased 3.1.0
1. Added an option to specify the JRE directory in the configuration file. 

Also update the readme.md with your new option and JSON config additions.

Add yourself to the AUTHORS file too (optional).

**Comment by karlsabo** [comment]
Oh, and to answer your question about building the packr-all JAR, yes, you have to run `gradlew PackrLauncher:publish` on every platform so the executable files are available. I use GitLab CI to do it all for me, you can look at `.gitlab-ci.yml` in the repository.

File: PackrLauncher/src/main/headers/packr.h (hunk 2 above)
Comment (petoncle) [reply PackrLauncher/src/main/headers/packr.h:44]: Yeah you're right I should replace string with dropt_char*. But to do that I need to find a way to 1. remove potential trailing slash from a dropt_char* (needed in packr.cpp), and 2. concatenate two dropt_char* (needed in packr_win32.cpp and packr_linux.cpp).

File: PackrLauncher/src/main/cpp/packr.cpp (hunk 4 above)
Comment (petoncle) [reply PackrLauncher/src/main/cpp/packr.cpp:504]: It is assumed that the jrePath value (coming from config.json) only contains forward slashes `/`. I've mentioned that in the description of the --jrePath command line parameter.
Both Linux and Mac's loadJNIFunctions() work with `/`.
Windows's loadJNIFunctions() works with `\`, and that's why I replace all `/` with `\` in Windows's loadJNIFunctions().

(I guess we could accept `\` in config.json and replace them with `/` in packr.cpp launchJavaVM().)

Now to answer your question, a trailing `/` breaks Windows and Linux's loadJNIFunctions() because we do:
```c++
jrePath + "/lib/server/libjvm.so" // Linux.
jrePath + "\\bin" // Windows.
```
That's why I remove any trailing `/` in packr.cpp launchJavaVM().

File: PackrLauncher/src/main/headers/packr.h (hunk 2 above)
Comment (karlsabo) [reply PackrLauncher/src/main/headers/packr.h:44]: 1. There's an example of that in packr.cpp when calling `changeWorkingDir`, look at the `launchJavaVM` function.
2. See below, there are C string functions for this, will need to use the correct one on Windows.
3. There's similar code in packr.cpp see the top of the file for #ifdef UNICODE. Then look at: https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/strcat-wcscat-mbscat?view=msvc-160
   * ```cpp
     #ifdef UNICODE
     #include <codecvt>
     #define stringCompare wcscmp
     #define findLastCharacter wcsrchr
     #else
     #define stringCompare strcmp
     #define findLastCharacter strrchr
     #endif
     ```

File: PackrLauncher/src/main/cpp/win32/packr_win32.cpp (hunk 1 above)
Comment (petoncle) [reply PackrLauncher/src/main/cpp/win32/packr_win32.cpp:259]: The reason for changing is that I would get the following error:
```
packr_win32.cpp(295): error C2664: 'void loadLibraries(const PTCHAR)': 
cannot convert argument 1 from 'const dropt_char *' to 'const PTCHAR'
```

File: PackrLauncher/src/main/headers/packr.h (hunk 2 above)
Comment (petoncle) [reply PackrLauncher/src/main/headers/packr.h]: Thanks @karlsabo. I've made the changes to replace std::string with dropt_char*. Let me know if there if it looks fine.

**Review by karlsabo — CHANGES_REQUESTED** [review]
Code looks good, thanks for the PR.

Can you update the changelog and readme files?

**Comment by petoncle** [comment]
While reviewing the README I realized that you can pass a JSON as input to the packr tool instead of using command line arguments. I was missing code to parse the new jrePath option from that JSON and so I've added it just now.

**Review by karlsabo — APPROVED** [review]
Thanks @petoncle!
